src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,317 @@
+            }
+
+            value = cacheItem.Value;
+            _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+            return _map.TryRemove(key, out _);
+        }
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal LinkedList<LRUCacheItem<TKey, TValue>> LinkedList => _doubleLinkedList;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long LinkedListCount => _doubleLinkedList.Count;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long MapCount => _map.Count;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal ICollection<LRUCacheItem<TKey, TValue>> MapValues => _map.Values;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long EventQueueCount => _eventQueue.Count;
+
+        /// <summary>
+        /// Calls <see cref="Dispose(bool)"/> and <see cref="GC.SuppressFinalize"/>
+        /// </summary>
+        public void Dispose()
+        {
+            // Dispose of unmanaged resources.
+            Dispose(true);
+            // Suppress finalization.
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// If <paramref name="disposing"/> is true, this method disposes of <see cref="_eventQueue"/>.
+        /// </summary>
+        /// <param name="disposing">True if called from the <see cref="Dispose()"/> method, false otherwise.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                _disposed = true;
+                if (disposing)
+                {
+                    _eventQueue.Dispose();
+                }
+            }
+        }
+    }
+
+    internal class LRUCacheItem<TKey, TValue>
+    {
+        internal TKey Key { get; }
+        internal TValue Value { get; set; }
+        internal DateTime ExpirationTime { get; set; }
+
+        internal LRUCacheItem(TKey key, TValue value)
+        {
+            Key = key ?? throw LogHelper.LogArgumentNullException(nameof(key));
+            Value = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
+        }
+
+        internal LRUCacheItem(TKey key, TValue value, DateTime expirationTime)
+        {
+            Key = key ?? throw LogHelper.LogArgumentNullException(nameof(key));
+            Value = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
+            ExpirationTime = expirationTime;
+        }
+
+        public override bool Equals(object obj)
+        {
+            LRUCacheItem<TKey, TValue> item = obj as LRUCacheItem<TKey, TValue>;
+            return item == null ? false : Key.Equals(item.Key);
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```

Comment:
this will use objects GetHashCode, this is undesirable. Hashcode should be stable for the lifetime of an object, so probably a seed value combined with the HashCode of Key

Reply:
Thanks for pointing this out. I've addressed the issue here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1605/commits/af3243229ca3f4885398299f0d733066d3b9be46.